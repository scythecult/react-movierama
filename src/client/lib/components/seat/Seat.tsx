import clsx from 'clsx';
import styles from './styles.module.css';

type SeatProps = {
  onClick?: () => void;
  isSelected?: boolean;
  isOccupied?: boolean;
  isDisabled?: boolean;
};

export const Seat = (props: SeatProps) => {
  const { onClick, isSelected, isOccupied, isDisabled = false } = props;
  const classNameFinal = clsx(styles.seat, {
    [styles.selected]: isSelected,
    [styles.occupied]: isOccupied,
    [styles.disabled]: isDisabled,
  });

  return <button className={classNameFinal} onClick={onClick} type="button" disabled={isOccupied}></button>;
};
