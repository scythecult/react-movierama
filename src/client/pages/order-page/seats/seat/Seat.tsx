import clsx from 'clsx';
import { SEAT_STATE, SEAT_TYPE, SeatStateMap } from '@/client/lib/constants/common';
import type { PropsWithClassName } from '@/client/lib/types/PropsWithClassName';
import styles from './styles.module.css';

export type SeatProps = PropsWithClassName & {
  onClick?: () => void;
  x?: number;
  y?: number;
  w?: number;
  h?: number;
  type?: number;
  place?: number;
  state?: number;
};

export const Seat = (props: SeatProps) => {
  const { x, y, w, h, type = 1, state = 1, place, className, onClick } = props;
  const seatType = SEAT_TYPE[type];
  const seatState = SEAT_STATE[state];
  const isOccupied = seatState === SEAT_STATE[SeatStateMap.OCCUPIED];
  const isSelected = seatState === SEAT_STATE[SeatStateMap.SELECTED];
  const classNameFinal = clsx(styles.seat, className, styles[seatType], styles[seatState]);

  return (
    <button
      className={classNameFinal}
      style={{
        left: x,
        top: y,
        width: w,
        height: h,
      }}
      onClick={onClick}
      type="button"
      disabled={isOccupied}
      data-test-id="seat"
      data-test-disabled={isOccupied}
      data-test-selected={isSelected}
    >
      <span className={styles.seatPlace} data-test-id="seat-place">
        {place}
      </span>
      <svg
        className={styles.seatIcon}
        fill="currentColor"
        width="18"
        height="20"
        viewBox="0 0 18 20"
        xmlns="http://www.w3.org/2000/svg"
        data-test-id="seat-icon"
      >
        <path d="M16.813 0c-.434 0-1.071.578-1.33.828l-.623.611a.572.572 0 0 1-.402.16.574.574 0 0 1-.402-.16l-.633-.613C12.998.416 12.293 0 11.769 0s-1.229.416-1.654.826l-.634.613a.578.578 0 0 1-.405.16.578.578 0 0 1-.405-.16L8.036.826C7.622.426 6.901 0 6.382 0c-.52 0-1.24.426-1.654.826l-.635.613a.579.579 0 0 1-.405.16.579.579 0 0 1-.405-.16L2.648.826C2.613.793 1.772 0 1.263 0 .683 0 .461.787.461 1.461V18.54c0 .674.221 1.461.802 1.461.49 0 1.279-.742 1.365-.825l.645-.614a.585.585 0 0 1 .409-.16c.155 0 .3.057.408.16l.636.613c.414.4 1.135.826 1.655.826s1.24-.426 1.655-.826l.635-.613a.578.578 0 0 1 .405-.16c.155 0 .299.057.405.16l.635.613c.414.4 1.135.826 1.654.826.52 0 1.24-.426 1.655-.826l.634-.613a.578.578 0 0 1 .405-.16c.155 0 .299.057.405.16l.635.613c.32.308.9.826 1.309.826.478 0 .726-.519.726-1.461V1.46c0-.94-.249-1.46-.726-1.46zm-3.018 15.023H9.593v-1.735h4.2v1.735zm0-4.201H4.296V9.178h9.497v1.644zm0-4.11H4.296V4.977h9.497v1.735z"></path>
      </svg>
    </button>
  );
};
