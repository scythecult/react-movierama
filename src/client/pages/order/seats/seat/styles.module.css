.seat {
    position: relative;
    display: flex;
    justify-content: center;
    align-items: center;
    width: 40px;
    height: 34px;
    padding: 4px;
    color: var(--app-seat-text);
    border-top-left-radius: var(--app-border-radius-m);
    border-top-right-radius: var(--app-border-radius-m);
    background-color: var(--app-seat-regular-color);
    border: 1px solid transparent;
    transition-property: scale, background-color, border-color, color, opacity;
    transition-duration: 0.2s;
    cursor: pointer;
}

.seat.selected {
    color: var(--app-secondary-text);
    background-color: var(--app-cyan-color);
}

.seat.occupied {
    background-color: var(--app-white-color);
}

.seat:hover.seat.occupied {
    border-color: transparent;
    scale: 1;
    cursor: not-allowed;
}

.seat:hover {
    border-color: var(--app-blue-color);
    scale: 1.2;
    cursor: pointer;
}

.seat:hover .seatPlace {
    opacity: 0;
    pointer-events: none;
}

.seat:hover .seatIcon {
    opacity: 1;
}

.seat:focus,
.seat:focus-visible {
    outline: 4px auto -webkit-focus-ring-color;
}

.seat.disabled:hover {
    scale: 1;
    cursor: default;
}

.seat.selected:hover .seatIcon {
    top: 50%;
    left: 0;
    transform: translate(-50%, -50%);
}

.seat.selected:hover .seatPlace {
    opacity: 1;
}

.regular {
    background-color: var(--app-seat-regular-color);
}

.comfort {
    background-color: var(--app-seat-comfort-color);
}

.vip {
    background-color: var(--app-seat-vip-color);
}

.seatIcon {
    position: absolute;
    top: 50%;
    left: 50%;
    width: 18px;
    height: 20px;
    color: var(--app-white-color);
    transform: translate(-50%, -50%);
    transition-property: opacity, left, transform;
    transition-duration: 0.2s;
    opacity: 0;
    pointer-events: none;
}

