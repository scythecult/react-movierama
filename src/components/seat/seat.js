import React from "react";
import { useDispatch } from "react-redux";
import { toggleActive } from "../../features/seats/seats-slice";

import classes from "./seat.module.css";

const Seat = ({ id, isSelected, isOccupied, isDisabled = false }) => {
  const dispatch = useDispatch();

  let modifierClass = "";

  if (isSelected) {
    modifierClass = classes.selected;
  }

  if (isOccupied) {
    modifierClass = `${classes.occupied} ${classes.disabled}`;
  }

  if (isDisabled) {
    modifierClass += ` ${classes.disabled}`;
  }

  const handleClick = () => {
    dispatch(toggleActive({ id }));
  };

  return <div className={`${classes.seat} ${modifierClass}`} onClick={handleClick}></div>;
};

export { Seat };
