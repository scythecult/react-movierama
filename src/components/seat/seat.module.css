.seat {
  width: 15px;
  height: 12px;
  margin: 3px;
  border-top-left-radius: 10px;
  border-top-right-radius: 10px;
  background-color: #444451;
}

.seat.selected {
  background-color: #6feaf6;
}

.seat.occupied {
  background-color: #ffffff;
}

/* .seat:nth-of-type(2) {
  margin-right: 18px;
}

.seat:nth-last-of-type(2) {
  margin-left: 18px;
} */

.seat:hover {
  cursor: pointer;
  transform: scale(1.2);
}

.seat.disabled:hover {
  cursor: default;
  transform: scale(1);
}
