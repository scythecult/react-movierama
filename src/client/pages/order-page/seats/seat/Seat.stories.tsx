import type { Meta, StoryFn } from '@storybook/react-vite';
import { Seat, type SeatProps } from './Seat';

export default {
  title: 'Pages/Seat',
  component: Seat,
  args: {
    place: 1,
  },
} satisfies Meta<SeatProps>;

const Template: StoryFn<SeatProps> = (props) => <Seat {...props} />;
export const Default = Template.bind(null);

export const Comfort = Template.bind(null);
Comfort.args = {
  type: 2,
};

export const Vip = Template.bind(null);
Vip.args = {
  type: 3,
};

export const Wheelchair = Template.bind(null);
Wheelchair.args = {
  type: 4,
};

export const Occupied = Template.bind(null);
Occupied.args = {
  state: 2,
};

export const Selected = Template.bind(null);
Selected.args = {
  state: 3,
};
