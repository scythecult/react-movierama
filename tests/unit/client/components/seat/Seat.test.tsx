import { render } from '@testing-library/react';
import { Seat } from '@/client/lib/components/seat/Seat';

describe('Seat', () => {
  test('should correspond default layout', () => {
    const result = render(<Seat />);

    expect(result.container).toMatchSnapshot();
  });

  test('should support the "isSelected" prop', () => {
    let result = render(<Seat isSelected />);

    expect(result.container).toMatchSnapshot();

    result = render(<Seat isSelected={false} />);

    expect(result.container).toMatchSnapshot();
  });

  test('should support the "isOccupied" prop', () => {
    let result = render(<Seat isOccupied />);

    expect(result.container).toMatchSnapshot();

    result = render(<Seat isOccupied={false} />);

    expect(result.container).toMatchSnapshot();
  });

  test('should support the "isDisabled" prop', () => {
    let result = render(<Seat isDisabled />);

    expect(result.container).toMatchSnapshot();

    result = render(<Seat isDisabled={false} />);

    expect(result.container).toMatchSnapshot();
  });
});
