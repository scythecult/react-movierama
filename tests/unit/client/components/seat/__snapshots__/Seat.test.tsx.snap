// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Seat > should correspond default layout 1`] = `
<div>
  <button
    class="_seat_357f74"
    type="button"
  />
</div>
`;

exports[`Seat > should support the "isDisabled" prop 1`] = `
<div>
  <button
    class="_seat_357f74 _disabled_357f74"
    type="button"
  />
</div>
`;

exports[`Seat > should support the "isDisabled" prop 2`] = `
<div>
  <button
    class="_seat_357f74"
    type="button"
  />
</div>
`;

exports[`Seat > should support the "isOccupied" prop 1`] = `
<div>
  <button
    class="_seat_357f74 _occupied_357f74"
    disabled=""
    type="button"
  />
</div>
`;

exports[`Seat > should support the "isOccupied" prop 2`] = `
<div>
  <button
    class="_seat_357f74"
    type="button"
  />
</div>
`;

exports[`Seat > should support the "isSelected" prop 1`] = `
<div>
  <button
    class="_seat_357f74 _selected_357f74"
    type="button"
  />
</div>
`;

exports[`Seat > should support the "isSelected" prop 2`] = `
<div>
  <button
    class="_seat_357f74"
    type="button"
  />
</div>
`;
